tests/test_ec2/test_instances.py

```diff
@@ -3044,6 +3044,58 @@ def test_create_instance_with_launch_template_id_produces_no_warning(
     assert len(captured_warnings) == 0
 
 
+@mock_ec2
+def test_run_instance_and_associate_public_ip():
+    ec2 = boto3.resource("ec2", "us-west-1")
+    vpc = ec2.create_vpc(CidrBlock="10.0.0.0/16")
+    subnet = ec2.create_subnet(VpcId=vpc.id, CidrBlock="10.0.0.0/18")
+
+    # Do not pass AssociatePublicIpAddress-argument
+    instance = ec2.create_instances(
+        ImageId=EXAMPLE_AMI_ID,
+        MinCount=1,
+        MaxCount=1,
+        NetworkInterfaces=[{"DeviceIndex": 0}],
+        SubnetId=subnet.id,
+    )[0]
+    interfaces = instance.network_interfaces_attribute
+    addresses = interfaces[0]["PrivateIpAddresses"][0]
+    addresses.should.have.key("Primary").equal(True)
+    addresses.should.have.key("PrivateIpAddress")
+    addresses.shouldnt.have.key("Association")
+
+    # Pass AssociatePublicIpAddress=False
+    instance = ec2.create_instances(
+        ImageId=EXAMPLE_AMI_ID,
+        MinCount=1,
+        MaxCount=1,
+        NetworkInterfaces=[{"DeviceIndex": 0, "AssociatePublicIpAddress": False}],
+        SubnetId=subnet.id,
+    )[0]
+    interfaces = instance.network_interfaces_attribute
+    addresses = interfaces[0]["PrivateIpAddresses"][0]
+    addresses.should.have.key("Primary").equal(True)
+    addresses.should.have.key("PrivateIpAddress")
+    addresses.shouldnt.have.key("Association")
```

Comment:
Shouldn't this check that there is no public IP address (or PublicDnsName)?

Reply:
I just verified this against AWS, and this is the response from the `describe_network_interfaces` call:
```
"PrivateIpAddresses": [{
    "Primary": true,
    "PrivateIpAddress": "10.0.33.93"
}]
```
AWS doesn't even have the 'Association'-key, let alone `Association.PublicIp`. So this assert looks correct to me.

Only when you pass `AssociatePublicIpAddress=True` does it return the `Association`-dict:
```
"PrivateIpAddresses": [{
    "Association": {
        "IpOwnerId": "amazon",
        "PublicDnsName": "",
        "PublicIp": "54.153.xx.yyy"
    },
    "Primary": true,
    "PrivateIpAddress": "10.0.62.61"
}]
```

Reply:
Ah gotcha thanks for the clarification!